contracts/src/v0.8/dev/keeper2_0/KeeperRegistry2_0.sol

```diff
@@ -541,26 +583,467 @@ contract KeeperRegistry2_0 is
+    uint8[] memory signerIndices = new uint8[](rs.length);
+    // Verify signatures attached to report
+    {
+      bytes32 h = keccak256(abi.encodePacked(keccak256(report), reportContext));
+      // i-th byte counts number of sigs made by i-th signer
+      uint256 signedCount = 0;
+
+      Signer memory signer;
+      address signerAddress;
+      for (uint256 i = 0; i < rs.length; i++) {
+        signerAddress = ecrecover(h, uint8(rawVs[i]) + 27, rs[i], ss[i]);
+        signer = s_signers[signerAddress];
+        if (!signer.active) revert OnlyActiveSigners();
+        unchecked {
+          signedCount += 1 << (8 * signer.index);
+        }
+        signerIndices[i] = signer.index;
+      }
+
+      if (signedCount & ORACLE_MASK != signedCount) revert DuplicateSigners();
+    }
+    return signerIndices;
+  }
+
+  /**
+   * @dev calls the Upkeep target with the performData param passed in by the
+   * transmitter and the exact gas required by the Upkeep
+   */
+  function _performUpkeep(Upkeep memory upkeep, bytes memory performData)
+    private
+    nonReentrant
+    returns (bool success, uint256 gasUsed)
+  {
+    gasUsed = gasleft();
+    bytes memory callData = abi.encodeWithSelector(PERFORM_SELECTOR, performData);
+    success = _callWithExactGas(upkeep.executeGas, upkeep.target, callData);
     gasUsed = gasUsed - gasleft();
-    uint96 payment = _calculatePaymentAmount(gasUsed, params.fastGasWei, params.linkEth, true);
 
-    s_upkeep[params.id].balance = s_upkeep[params.id].balance - payment;
-    s_upkeep[params.id].amountSpent = s_upkeep[params.id].amountSpent + payment;
-    s_upkeep[params.id].lastKeeper = params.from;
-    s_keeperInfo[params.from].balance = s_keeperInfo[params.from].balance + payment;
+    return (success, gasUsed);
+  }
 
-    emit UpkeepPerformed(params.id, success, params.from, payment, params.performData);
-    return success;
+  /**
+   * @dev does postPerform payment processing for an upkeep. Stores lastPerformBlock Calculates
+   * gasPayment and premiumPerSigner. Deducts upkeep's balance for the total payment
+   */
+  function _postPerformUpkeep(
+    HotVars memory hotVars,
+    uint256 upkeepId,
+    uint96 numSigners,
+    uint32 checkBlockNumber,
+    uint256 gasOverhead,
+    UpkeepTransmitInfo memory upkeepTransmitInfo
+  ) internal returns (uint96, uint96) {
+    s_upkeep[upkeepId].lastPerformBlockNumber = uint32(block.number);
+    (uint96 gasPayment, uint96 premium) = _calculatePaymentAmount(
+      hotVars,
+      upkeepTransmitInfo.gasUsed,
+      gasOverhead,
+      upkeepTransmitInfo.paymentParams.fastGasWei,
+      upkeepTransmitInfo.paymentParams.linkNative,
+      true
+    );
+    uint96 premiumPerSigner = premium / numSigners;
+    uint96 total = gasPayment + premiumPerSigner * numSigners;
+
+    s_upkeep[upkeepId].balance -= total;
+    s_upkeep[upkeepId].amountSpent += total;
+
+    UpkeepPerformedLogFields memory upkeepPerformedLogFields = UpkeepPerformedLogFields({
```

Comment:
Added this as I could not log all these fields due to stack pressure. I feel all these fields are useful info which should be logged but can trim them to simplify.

Reply:
FYI, this could end up being quite pricey

Reply:
removed couple of params and ended up not needing struct